# 9. The full S-expansion pipeline: $\mathfrak{sl}(2)$ expanded with $S^{991}_{(5)}$

This is the package's flagship example, porting the Java example `II_SExp_sl2_S991` from
Sections 5.5–5.6 of the paper (C. Inostroza, I. Kondrashuk, N. Merino, F. Nadal, *A Java Library to Perform S-Expansions of Lie Algebras*, Axioms (2025)). Starting from $\mathfrak{sl}(2, \mathbb{R})$ and
the order-5 catalogued semigroup $S^{991}_{(5)}$, we compute all four algebras the
S-expansion method produces:

| Algebra | Construction | Generators |
|---|---|---|
| $G_S$ | expanded algebra $S \otimes G$ | 15 |
| $G_{S,R}$ | resonant subalgebra of $G_S$ | 9 |
| $G_{S,\mathrm{red}}$ | $0_S$-reduction of $G_S$ | 12 |
| $G_{S,R,\mathrm{red}}$ | $0_S$-reduction of $G_{S,R}$ | 6 |

For each algebra we print its non-vanishing commutators, structure constants and Cartan–Killing
metric, with the determinant and eigenvalue signature of the metric. All four algebras turn out to
be semi-simple ($\det \neq 0$), and the final 6-generator algebra is
$\mathfrak{su}(2) \oplus \mathfrak{sl}(2, \mathbb{R})$ with signature $(2, 4, 0)$ —
the result highlighted in Section 5.6 of the paper.

In [1]:
from sexpansion import LieAlgebra, Resonance, load_semigroup
from sexpansion.reports import commutator_table, metric_report, structure_constants_report

## Setup

Load $\mathfrak{sl}(2)$ (basis of notebook 8) and the semigroup $S^{991}_{(5)}$ from the bundled
catalogue. The semigroup has a zero element — $\lambda_1$, index 0 — which is what makes the
$0_S$-reductions below possible.

In [2]:
sl2 = LieAlgebra.sl2()
s991 = load_semigroup(5, 991)

print("Semigroup S991 (0-based table):")
print(s991)
print("Zero element:", s991.find_zero())

Semigroup S991 (0-based table):
0 0 0 0 0
0 1 2 3 4
0 2 1 4 3
0 3 4 2 1
0 4 3 1 2
Zero element: 0


## Resonant decomposition

The paper decomposes $S = S_0 \cup S_1$ with $S_0 = \{\lambda_1, \lambda_2, \lambda_3\}$
and $S_1 = \{\lambda_1, \lambda_4, \lambda_5\}$, and grades the algebra as
$G = V_0 \oplus V_1$ with $V_0 = \mathrm{span}\{X_1\}$ (the maximal abelian subalgebra found
in notebook 8) and $V_1 = \mathrm{span}\{X_2, X_3\}$. The code uses 0-based indices
throughout, so the paper's 1-based labels shift down by one.

In [3]:
resonance = Resonance(s0=(0, 1, 2), s1=(0, 3, 4))
v0, v1 = (0,), (1, 2)

## Compute the four algebras

`s_expand` builds $G_S = S \otimes G$; `resonant_subalgebra` keeps the generators in
$(S_0 \otimes V_0) \cup (S_1 \otimes V_1)$; `zero_reduced` discards the generators attached to
the zero element. The helper below prints the full report for one algebra — generators are
labelled $X_{(i,\alpha)}$, where $i$ is the $\mathfrak{sl}(2)$ index and $\alpha$ the
semigroup element (both 1-based).

In [4]:
expanded = sl2.s_expand(s991)
resonant = expanded.resonant_subalgebra(resonance, v0, v1)
reduced = expanded.zero_reduced()
resonant_reduced = resonant.zero_reduced()


def describe(algebra):
    print(f"{algebra.n_generators} generators")
    print("\nNon-vanishing commutators:")
    print(commutator_table(algebra, one_based=True))
    print("\nNon-vanishing structure constants:")
    print(structure_constants_report(algebra, one_based=True))
    print("\nCartan-Killing metric:")
    print(metric_report(algebra.cartan_killing_metric(), precision=0))
    print(f"det = {algebra.det():.6g}   signature (+,-,0) = {algebra.signature()}")

## The expanded algebra $G_S$

The full tensor product: $3 \times 5 = 15$ generators.

In [5]:
describe(expanded)

15 generators

Non-vanishing commutators:
[X_(1,1), X_(2,1)] = -2 X_(3,1)
[X_(1,1), X_(2,2)] = -2 X_(3,1)
[X_(1,1), X_(2,3)] = -2 X_(3,1)
[X_(1,1), X_(2,4)] = -2 X_(3,1)
[X_(1,1), X_(2,5)] = -2 X_(3,1)
[X_(1,1), X_(3,1)] = 2 X_(2,1)
[X_(1,1), X_(3,2)] = 2 X_(2,1)
[X_(1,1), X_(3,3)] = 2 X_(2,1)
[X_(1,1), X_(3,4)] = 2 X_(2,1)
[X_(1,1), X_(3,5)] = 2 X_(2,1)
[X_(1,2), X_(2,1)] = -2 X_(3,1)
[X_(1,2), X_(2,2)] = -2 X_(3,2)
[X_(1,2), X_(2,3)] = -2 X_(3,3)
[X_(1,2), X_(2,4)] = -2 X_(3,4)
[X_(1,2), X_(2,5)] = -2 X_(3,5)
[X_(1,2), X_(3,1)] = 2 X_(2,1)
[X_(1,2), X_(3,2)] = 2 X_(2,2)
[X_(1,2), X_(3,3)] = 2 X_(2,3)
[X_(1,2), X_(3,4)] = 2 X_(2,4)
[X_(1,2), X_(3,5)] = 2 X_(2,5)
[X_(1,3), X_(2,1)] = -2 X_(3,1)
[X_(1,3), X_(2,2)] = -2 X_(3,3)
[X_(1,3), X_(2,3)] = -2 X_(3,2)
[X_(1,3), X_(2,4)] = -2 X_(3,5)
[X_(1,3), X_(2,5)] = -2 X_(3,4)
[X_(1,3), X_(3,1)] = 2 X_(2,1)
[X_(1,3), X_(3,2)] = 2 X_(2,3)
[X_(1,3), X_(3,3)] = 2 X_(2,2)
[X_(1,3), X_(3,4)] = 2 X_(2,5)
[X_(1,3), X_(3,5)] = 2 X_(2,4)
[X_(1,4), X_(

## The resonant subalgebra $G_{S,R}$

Keeping $(S_0 \otimes V_0) \cup (S_1 \otimes V_1)$ leaves $3 \cdot 1 + 3 \cdot 2 = 9$
generators. The resonance conditions guarantee this subset closes as a Lie algebra.

In [6]:
describe(resonant)

9 generators

Non-vanishing commutators:
[X_(1,1), X_(2,1)] = -2 X_(3,1)
[X_(1,1), X_(2,4)] = -2 X_(3,1)
[X_(1,1), X_(2,5)] = -2 X_(3,1)
[X_(1,1), X_(3,1)] = 2 X_(2,1)
[X_(1,1), X_(3,4)] = 2 X_(2,1)
[X_(1,1), X_(3,5)] = 2 X_(2,1)
[X_(1,2), X_(2,1)] = -2 X_(3,1)
[X_(1,2), X_(2,4)] = -2 X_(3,4)
[X_(1,2), X_(2,5)] = -2 X_(3,5)
[X_(1,2), X_(3,1)] = 2 X_(2,1)
[X_(1,2), X_(3,4)] = 2 X_(2,4)
[X_(1,2), X_(3,5)] = 2 X_(2,5)
[X_(1,3), X_(2,1)] = -2 X_(3,1)
[X_(1,3), X_(2,4)] = -2 X_(3,5)
[X_(1,3), X_(2,5)] = -2 X_(3,4)
[X_(1,3), X_(3,1)] = 2 X_(2,1)
[X_(1,3), X_(3,4)] = 2 X_(2,5)
[X_(1,3), X_(3,5)] = 2 X_(2,4)
[X_(2,1), X_(3,1)] = 2 X_(1,1)
[X_(2,1), X_(3,4)] = 2 X_(1,1)
[X_(2,1), X_(3,5)] = 2 X_(1,1)
[X_(2,4), X_(3,1)] = 2 X_(1,1)
[X_(2,4), X_(3,4)] = 2 X_(1,3)
[X_(2,4), X_(3,5)] = 2 X_(1,2)
[X_(2,5), X_(3,1)] = 2 X_(1,1)
[X_(2,5), X_(3,4)] = 2 X_(1,2)
[X_(2,5), X_(3,5)] = 2 X_(1,3)

Non-vanishing structure constants:
C_(1,1)(2,1)^(3,1) = -2
C_(1,1)(2,4)^(3,1) = -2
C_(1,1)(2,5)^(3,1) = -2
C_(1,

## The reduced algebra $G_{S,\mathrm{red}}$

The $0_S$-reduction of $G_S$ removes the three generators $X_{(i,1)}$ attached to the zero
element, leaving 12.

In [7]:
describe(reduced)

12 generators

Non-vanishing commutators:
[X_(1,2), X_(2,2)] = -2 X_(3,2)
[X_(1,2), X_(2,3)] = -2 X_(3,3)
[X_(1,2), X_(2,4)] = -2 X_(3,4)
[X_(1,2), X_(2,5)] = -2 X_(3,5)
[X_(1,2), X_(3,2)] = 2 X_(2,2)
[X_(1,2), X_(3,3)] = 2 X_(2,3)
[X_(1,2), X_(3,4)] = 2 X_(2,4)
[X_(1,2), X_(3,5)] = 2 X_(2,5)
[X_(1,3), X_(2,2)] = -2 X_(3,3)
[X_(1,3), X_(2,3)] = -2 X_(3,2)
[X_(1,3), X_(2,4)] = -2 X_(3,5)
[X_(1,3), X_(2,5)] = -2 X_(3,4)
[X_(1,3), X_(3,2)] = 2 X_(2,3)
[X_(1,3), X_(3,3)] = 2 X_(2,2)
[X_(1,3), X_(3,4)] = 2 X_(2,5)
[X_(1,3), X_(3,5)] = 2 X_(2,4)
[X_(1,4), X_(2,2)] = -2 X_(3,4)
[X_(1,4), X_(2,3)] = -2 X_(3,5)
[X_(1,4), X_(2,4)] = -2 X_(3,3)
[X_(1,4), X_(2,5)] = -2 X_(3,2)
[X_(1,4), X_(3,2)] = 2 X_(2,4)
[X_(1,4), X_(3,3)] = 2 X_(2,5)
[X_(1,4), X_(3,4)] = 2 X_(2,3)
[X_(1,4), X_(3,5)] = 2 X_(2,2)
[X_(1,5), X_(2,2)] = -2 X_(3,5)
[X_(1,5), X_(2,3)] = -2 X_(3,4)
[X_(1,5), X_(2,4)] = -2 X_(3,2)
[X_(1,5), X_(2,5)] = -2 X_(3,3)
[X_(1,5), X_(3,2)] = 2 X_(2,5)
[X_(1,5), X_(3,3)] = 2 X_(2,4)
[X_(1,5), X_

## The reduction of the resonant subalgebra $G_{S,R,\mathrm{red}}$

Applying the $0_S$-reduction to $G_{S,R}$ leaves 6 generators.

In [8]:
describe(resonant_reduced)

6 generators

Non-vanishing commutators:
[X_(1,2), X_(2,4)] = -2 X_(3,4)
[X_(1,2), X_(2,5)] = -2 X_(3,5)
[X_(1,2), X_(3,4)] = 2 X_(2,4)
[X_(1,2), X_(3,5)] = 2 X_(2,5)
[X_(1,3), X_(2,4)] = -2 X_(3,5)
[X_(1,3), X_(2,5)] = -2 X_(3,4)
[X_(1,3), X_(3,4)] = 2 X_(2,5)
[X_(1,3), X_(3,5)] = 2 X_(2,4)
[X_(2,4), X_(3,4)] = 2 X_(1,3)
[X_(2,4), X_(3,5)] = 2 X_(1,2)
[X_(2,5), X_(3,4)] = 2 X_(1,2)
[X_(2,5), X_(3,5)] = 2 X_(1,3)

Non-vanishing structure constants:
C_(1,2)(2,4)^(3,4) = -2
C_(1,2)(2,5)^(3,5) = -2
C_(1,2)(3,4)^(2,4) = 2
C_(1,2)(3,5)^(2,5) = 2
C_(1,3)(2,4)^(3,5) = -2
C_(1,3)(2,5)^(3,4) = -2
C_(1,3)(3,4)^(2,5) = 2
C_(1,3)(3,5)^(2,4) = 2
C_(2,4)(3,4)^(1,3) = 2
C_(2,4)(3,5)^(1,2) = 2
C_(2,5)(3,4)^(1,2) = 2
C_(2,5)(3,5)^(1,3) = 2

Cartan-Killing metric:
-16   0   0   0   0   0
  0 -16   0   0   0   0
  0   0   0  16   0   0
  0   0  16   0   0   0
  0   0   0   0   0  16
  0   0   0   0  16   0
det = 1.67772e+07   signature (+,-,0) = (2, 4, 0)


## Result

Every level of the pipeline preserves semi-simplicity:

| Algebra | det | signature $(+,-,0)$ |
|---|---|---|
| $G_S$ | $5.9 \times 10^{20}$ | $(9, 6, 0)$ |
| $G_{S,R}$ | $-8.6 \times 10^{9}$ | $(4, 5, 0)$ |
| $G_{S,\mathrm{red}}$ | $-1.2 \times 10^{18}$ | $(7, 5, 0)$ |
| $G_{S,R,\mathrm{red}}$ | $1.7 \times 10^{7}$ | $(2, 4, 0)$ |

The final algebra's Cartan–Killing metric splits into a negative-definite 3-generator block —
the compact form $\mathfrak{su}(2)$, signature $(0,3,0)$ — and a block of signature $(2,1,0)$ —
the split form $\mathfrak{sl}(2,\mathbb{R})$. The 6-generator resonant-reduced algebra is
therefore $\mathfrak{su}(2) \oplus \mathfrak{sl}(2, \mathbb{R})$, reproducing Section 5.6 of
the paper. This entire computation is the package's README quickstart.